### PR TITLE
feat(replays): Incremental render of replay details page

### PR DESCRIPTION
**Pull request by ryan953**
Fetch segments in parallel, and incrementally re-render whenever replay data (core record, segment or error data) arrives.

**Before:**
We could see a huge waterfall of requests like this for segment data. 21 sequential requests takes a long time:
SCR-20230213-ouz

**After:**
Now we're just hammering the network to fetch all the data. 
SCR-20230214-dex

The whole UI will re-render as data appears. The p95 for all replays is ~40 segments, so there will not be any incremental render because we'll get all the data at once!

When there is lots of data we'll see information pop in when it's ready, and the ReplayPlayer (the video area) will have a loading indicator until everything is ready. The progression looks like this:

(note this gif has fewer steps than what we'd see loading thie page for real, but it's illustrative of the stages that the page will go through)

incremental loading

### CODE REVIEW

**Review by billyvg — APPROVED** [review]
I wonder if we need a loading indicator that encompasses the pieces of the UI that are progressively loading. From the screenshot it only looks like the Player is loading, when it's really everything else as well.

Since we're generating the duration based on the event timestamps as well, does that 1) update and 2) be confusing for users if it does change?

Some other thoughts: for replays where # of segments < 100, this doesn't change much, and will still appear to be pretty slow if it's a large replay too. The initial checkout segment is where I'm guessing is going to be the largest so I wonder if we can break these requests up as well with less included segments in the first request.

File: static/app/utils/replays/hooks/useReplayData.tsx [hunk 1]
```diff
@@ -95,62 +107,44 @@ function useReplayData({replaySlug, orgSlug}: Options): Result {
...
+      await Promise.allSettled(
+        cursors.map(cursor => {
+          const promise = api.requestPromise(
+            `${baseUrl}?download&per_page=${perPage}&cursor=${cursor}`
+          );
+          promise.then(response => {
+            const attachments = responsesToAttachments([response]);
+            setState(prev => ({
+              ...prev,
+              attachments: (prev.attachments ?? []).concat(attachments),
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/utils/replays/hooks/useReplayData.tsx:133]: It should be sorted inside the ReplayReader class which we return from this hook.

- we call `spansFactory()` for sorted network & memory spans. both need to be sorted before rendering the Timeline or Memory chart. 
- `rrwebEventListFactory()` for sorted rrweb data, which def needs to be sorted for the player

breadcrumbs might need to be sorted though yeah. the breadcrumb component doesn't seem to be doing it

File: static/app/utils/replays/hooks/useReplayData.tsx [hunk 2]
```diff
@@ -63,16 +61,30 @@ const INITIAL_STATE: State = Object.freeze({
   replayRecord: undefined,
 });
 
+function responsesToAttachments(responses: Array<unknown>) {
+  // Each response returns an array of segments
+  const segments = responses.flatMap(_ => _);
```
Comment (scttcper) [inline static/app/utils/replays/hooks/useReplayData.tsx:66]: is this different from `.flat()`

Comment (ryan953) [reply static/app/utils/replays/hooks/useReplayData.tsx:66]: d'oh

**Pull request merged**